design/external-references.md

```diff
@@ -0,0 +1,394 @@
+}
+```
+
+The `exportString` method creates an AWS CloudFormation Output for this value assigned to this export name.
+
+The `importString` method returns the value for this specific export name. When importing a value, users can specify the following `ImportOptions`:
+
+* If `type` is set to `Deployment`, the method will return an `Fn::ImportValue(exportName)` token. This means that CDK code cannot reason about the concrete value, which will only be resolved when the stack is deployed.
+* If `type` is set to `Synthesis` (default) the method will exercise an environmental context provider to look up the export value __during synthesis__. The concrete value will be propagated to the CDK app and can be reasoned about like any normal value.
+* The `weak` option is only relevant for synth-time resolution. If it is `false` (which is the default), the CDK will automatically embed a `Metadata` entry on the consuming resource with an `Fn::ImportValue`. This will force CloudFormation to take a strong reference on the export, even through the actual value is concretely resolved during synthesis. This ensures, for example, that the producing stack can't be deleted as long as there stacks consuming the exported values. This behavior (which is the default), can be disabled by settings `weak: true`, in which case the `Fn::ImportValue` will simply not be included.
+
+On top of these two methods, we can now define import and export methods for serializable objects:
+
+```ts
+class Stack {
+  exportObject(exportName: string, obj: ISerializable): void;
+  importObject(exportName: string, options?: ImportOptions): IDeserializationContext;
+}
+```
+
+The `exportObject` method will invoke `obj.serialize` with a serialization context "bound" to this export name. This means the export name will be used as a prefix to all written keys. `writeObject` will be implemented with a nested serialization context that adds another component to the export name prefix.
+
+The `importObject` method will be used like this:
+
+```ts
+const importedBucket = Bucket.deserializeBucket(stack.importObject('MyBucketExportName'));
+```
+
+The method will return a deserialization context that's bound to the export name. Similarly to the serialization context, it will prefix all values read through `readString` with the export name, and so forth with `readObject`.
+
+As mentioned about, the default resolve type for imports is `Synthesis` (with strong-references). This means that the values returned by `readString` will be actual concrete values. If users opt-in to deploy-time resolution (by setting using `Deployment` resolve type), the values returned will be tokens. In some cases this would be fine, but there could be constructs that cannot deal with opaque values (i.e. if the value is an list of strings and needs to be deconstructed). In those cases (REQ8), constructs should invoke `readString` with `allowUnresolved: false` to indicate that this specific value cannot be a token.
```

Comment:
I think the default behavior should be `allowUnresolved: false`, and people should opt into allowing tokens. The fact everything is a `string` makes it all too  easy for people to forget to check for unresolved values... I think not allowing by default is a safer behavior.